fs/btrfs/compression.c

```diff
@@ -1595,12 +1606,26 @@ int btrfs_compress_heuristic(struct inode *inode, u64 start, u64 end)
 
 unsigned int btrfs_compress_str2level(const char *str)
 {
-	if (strncmp(str, "zlib", 4) != 0)
-		return 0;
+	int ret;
+	unsigned int level;
+	unsigned int default_level = 0;
+	unsigned int max_level = 0;
+
+	if (strncmp(str, "zlib", 4) == 0) {
+		default_level = BTRFS_ZLIB_DEFAULT_LEVEL;
+		max_level = BTRFS_ZLIB_MAX_LEVEL;
+	}
 
-	/* Accepted form: zlib:1 up to zlib:9 and nothing left after the number */
-	if (str[4] == ':' && '1' <= str[5] && str[5] <= '9' && str[6] == 0)
-		return str[5] - '0';
+	if (strncmp(str, "zstd", 4) == 0) {
+		default_level = BTRFS_ZSTD_DEFAULT_LEVEL;
+		max_level = BTRFS_ZSTD_MAX_LEVEL;
+	}
```

Comment:
If someone adds a new caller that is neither zlib or zstd, the `str[4] == ':'` could trigger an out of bounds read. We should return 0 if it is neither.